### ISSUE TITLE
Comparison with the original

### ISSUE THREAD

**Issue by kayuksel** [zhang943/Approximate-HetConv#1]
Can you please compare this approximation to original? The advantages and disadvantages? 
In other words, what would be the trade-off of using your version. Here is the original one:

https://github.com/sxpro/HetConvolution2d_pytorch/

**Comment by zhang943** [comment]
In the original paper, figure3 shown the details of HetConv filters at a layer. Here I use a table to represent it. Each column is a filter and contails M channels/kernels.
<table>
<tr><td></td><td colspan="9" align="center">HetConv Filters (N)</td></tr>
    <tr><td rowspan="9" width="1">Kernels (M)</td><td>3x3</td><td>1x1</td><td>1x1</td><td>1x1</td><td>...</td><td>3x3</td><td>1x1</td><td>1x1</td><td>1x1</td></tr>
    <tr><td>1x1</td><td>3x3</td><td>1x1</td><td>1x1</td><td>...</td><td>1x1</td><td>3x3</td><td>1x1</td><td>1x1</td></tr>
    <tr><td>1x1</td><td>1x1</td><td>3x3</td><td>1x1</td><td>...</td><td>1x1</td><td>1x1</td><td>3x3</td><td>1x1</td></tr>
    <tr><td>1x1</td><td>1x1</td><td>1x1</td><td>3x3</td><td>...</td><td>1x1</td><td>1x1</td><td>1x1</td><td>3x3</td></tr>
    <tr><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td><td>...</td></tr>
    <tr><td>3x3</td><td>1x1</td><td>1x1</td><td>1x1</td><td>...</td><td>3x3</td><td>1x1</td><td>1x1</td><td>1x1</td></tr>
    <tr><td>1x1</td><td>3x3</td><td>1x1</td><td>1x1</td><td>...</td><td>1x1</td><td>3x3</td><td>1x1</td><td>1x1</td></tr>
    <tr><td>1x1</td><td>1x1</td><td>3x3</td><td>1x1</td><td>...</td><td>1x1</td><td>1x1</td><td>3x3</td><td>1x1</td></tr>
    <tr><td>1x1</td><td>1x1</td><td>1x1</td><td>3x3</td><td>...</td><td>1x1</td><td>1x1</td><td>1x1</td><td>3x3</td></tr>
</table>

In my view, using successive 3x3 kernels in a filter is equivalent to using non-successive ones. So the HetConv can be converted to the one below.
<table>
<tr><td></td><td colspan="9" align="center">HetConv Filters (N)</td></tr>
    <tr><td rowspan="9" width="1">Kernels (M)</td><td>3x3</td><td>1x1</td><td>1x1</td><td>1x1</td><td>...</td><td>3x3</td><td>1x1</td><td>1x1</td><td>1x1</td></tr>
    <tr><td>3x3</td><td>1x1</td><td>1x1</td><td>1x1</td><td>...</td><td>3x3</td><td>1x1</td><td>1x1</td><td>1x1</td></tr>
    <tr><td>1x1</td><td>3x3</td><td>1x1</td><td>1x1</td><td>...</td><td>1x1</td><td>3x3</td><td>1x1</td><td>1x1</td></tr>
    <tr><td>1x1</td><td>3x3</td><td>1x1</td><td>1x1</td><td>...</td><td>1x1</td><td>3x3</td><td>1x1</td><td>1x1</td></tr>
    <tr><td>1x1</td><td>1x1</td><td>3x3</td><td>1x1</td><td>...</td><td>1x1</td><td>1x1</td><td>3x3</td><td>1x1</td></tr>
    <tr><td>1x1</td><td>1x1</td><td>3x3</td><td>1x1</td><td>...</td><td>1x1</td><td>1x1</td><td>3x3</td><td>1x1</td></tr>
    <tr><td>1x1</td><td>1x1</td><td>1x1</td><td>3x3</td><td>...</td><td>1x1</td><td>1x1</td><td>1x1</td><td>3x3</td></tr>
    <tr><td>1x1</td><td>1x1</td><td>1x1</td><td>3x3</td><td>...</td><td>1x1</td><td>1x1</td><td>1x1</td><td>3x3</td></tr>
</table>

The output of each filter is the sum of the convolution of p\*M 3x3 kernels and (1-p)\*M 1x1 kernels.

I use group convolution and set group=p to compute the result of p\*M 3x3 kernels in each filter. And different with the original HetConv, I compute the 1x1 convolution on all M input channels rather than the specific (1-p)\*M ones. That's why I call it an approximation.

[sxpro](https://github.com/sxpro/HetConvolution2d_pytorch/) implements the original HetConv in an iterative manner, which i think may not be efficient.

**Comment by akaniklaus** [comment]
Thanks for the detailed explanation. Meanwhile, what do you think of the Octave convolution?

**Comment by zhang943** [comment]
@akaniklaus Sorry, I have not read this paper yet.

**Comment by irvinxav** [comment]
Thanks for the detailed explanation

This can be easily converted to original by making corresponding extra p*M 1x1 kernels weights to zero and masking the corresponding gradients so that extra p*M 1x1 kernels weights remains zero during backpropagations.

**Comment by irvinxav** [comment]
I have implemented original HetConv using same approach
https://github.com/irvinxav/Efficient-HetConv-Heterogeneous-Kernel-Based-Convolutions

**Comment by kayuksel** [comment]
I found the following approach in another repo, maybe it is also worth to try and compare.
```
        self.conv3x3 = nn.Conv2d(in_feats, out_feats, kernel_size=3, padding=1, groups=p)
        self.conv1x1_ = nn.Conv2d(in_feats, out_feats, kernel_size=1, groups=p)
        self.conv1x1 = nn.Conv2d(in_feats, out_feats, kernel_size=1)

    def forward(self, x):
        return self.conv3x3(x) + self.conv1x1(x) - self.conv1x1_(x)
```
